Step3 배포 스크립트 작성하기 리뷰 요청드립니다!

ilmare-cbk opened the pull request:
안녕하세요 우테캠 Pro 5기 교육생 천보경입니다.
3단계 미션 리뷰 요청드립니다.
- 2단계 추가 미션 중 `설정 별도로 관리하기` 부분을 이번 미션 3단계 적용해보았습니다! (-> submodule에 변경사항이 생겼을 때 main project와 함께 갱신하는 부분은 좀 더 공부가 필요할 것 같습니다😢)
- nginx  upstream 설정의  public ip를 private ip로 변경하였습니다!

---
지난 미션과 이번 미션을 진행하면서 몇가지 궁금한 사항이 있어 질문 드려요!

질문1) hibernate의 ddl-auto validate에 대해서 테스트를 해보았습니다. 
`새로운 Entity(테이블)가 생성된 경우`, `기존 Entity에 필드가 추가 또는 제거 된 경우` 에 대해서는 `SchemaManagementException` 예외가 발생하는 것을 확인할 수 있었습니다. 그런데 `기존 필드(컬럼)에 대한 속성 값 변경 (예 length 추가 등)`에 대해서는 validate가 작동하지 않았습니다! 이 경우에 대해서는 validate가 작동하지 않는 것이 맞는건가요? (혹시나 제가 잘못된 학습을 한건 아닐까해서 한 번 여쭤봐요 ㅎㅎ)

질문2)  database username과 password를 bash alias로 설정하여 스크립트에서 jar 배포 시 매개변수로 넣어보고자 테스트를 진행해봤는데 잘 되지 않았습니다😢 스크립트에서 alias로 설정한 값을 불러올 수 있는 방법이 있는지 궁금합니다!

ilmare-cbk committed:
build(config): git submodule 를 활용한 properties 설정

ilmare-cbk committed:
docs(README.md): step3 요구사항 정리

ilmare-cbk committed:
docs(README.md): step3 요구사항 수정

- 스크립트 내용 수정

ilmare-cbk committed:
docs(README.md): step3 요구사항 수정

- 스크립트 내용 수정
- 소스변경 변경사항이 없더라도 실행 중인 프로세스가 없다면 deploy.sh 통해 프로세스를 실행시킬 수 있도록 수정

seokhongkim commented:
보경님 안녕하세요! 질문 남겨주신 부분에 대한 답변 먼저 드립니다.

1. [stackoverflow](https://stackoverflow.com/questions/30863012/hibernate-validation-on-server-startup-does-not-check-column-length)에 올라온 내용을 참고해보시면 도움이 될 것 같습니다. 실제로 org.hibernate.mapping.Table#validateColumns 의 구현을 보시면 type 과 컬럼이 없는 경우에 대한 체크만 있는 것을 보실 수 있습니다. :)
2. [askubuntu](https://askubuntu.com/questions/98782/how-to-run-an-alias-in-a-shell-script)에 올라온 내용을 참고하시면 shell script 내에서 alias 를 사용하실 수 있습니다.
alias 로 설정하는 코드를 `.bashrc` 등에 추가하신다면, alias 대신 환경변수로 등록하는 방법을 고민해보셔도 좋겠습니다. [spring boot external config](https://docs.spring.io/spring-boot/docs/1.5.6.RELEASE/reference/html/boot-features-external-config.html) 문서를 참고해보시면 10번에 OS environment variables 도 설정값으로 사용할 수 있는 것을 확인하실 수 있습니다. :)

seokhongkim commented on README.md:
```diff
@@ -132,4 +132,155 @@ npm run dev
+- `/home/ubuntu/nextstep/script` 경로에 스크립트가 있습니다.
+- `/home/ubuntu/nextstep/log` 경로에 `app.log` 파일로 로그를 기록하고 있습니다.
+- `setting.sh`
+    - 여러 스크립트에서 공통으로 사용하는 변수와 함수를 정의해두었습니다.
```
공통으로 사용할 수 있는 함수를 별도로 분리하신 것 좋습니다. 👍 
이 스크립트에 있는 내용도 배포할 때만 사용할 것 같기도 하지만, 배포에 대한 코드와 공통으로 사용할 수 있는 코드가 나뉘어 있어 코드를 읽을 때는 가독성이 좋았습니다. :)

seokhongkim commented on README.md:
```diff
@@ -132,4 +132,155 @@ npm run dev
+          echo ""
+          echo ">> Pull Request 🏃♂️ "
+          echo ""
+          cd ${PROJECT_PATH} && git pull origin ${BRANCH}
```
필요한 브랜치만 pull 받아 오시는 것 좋습니다. 👍

seokhongkim commented on README.md:
```diff
@@ -132,4 +132,155 @@ npm run dev
+  BRANCH=$1
+  
+  ## guide
+  guide() {
```
기능별로 함수를 잘 나눠서 작성해주셨고, 배포 단계마다 로그 메시지도 잘 남겨주셔서 단계별로 눈에 잘 들어왔네요! 👍

seokhongkim commented on README.md:
```diff
@@ -132,4 +132,155 @@ npm run dev
+  
+  ## stop process
+  stop_process() {
+      JAR_NAME=$(cd ${JAR_PATH} && find ./* -name "*jar" | cut -c 3-)
```
`find_jar`, `find_pid` 함수를 미리 만들어놓으신 것 같은데, 이 함수를 활용하지 않으신 이유가 있을까요?

ilmare-cbk replied:
이 부분은 제가 기능별 스크립트를  처음에 따로 만들었다가 다시 새로 스크립트를 짜면서 미처 수정하지 못한 부분입니다!
함수를 사용해서 조회할 수 있도록 수정해보겠습니다😄

seokhongkim commented on README.md:
```diff
@@ -132,4 +132,155 @@ npm run dev
+      fi
+  
+      cd ${PROJECT_PATH} && git fetch
+      master=$(cd ${PROJECT_PATH} && git rev-parse ${BRANCH})
```
`master` 보다는 `local` 이 조금 더 저장하는 값을 잘 설명해주는 이름일 것 같기도 하네요! :)

ilmare-cbk replied:
넵! 스크립트도 명확한 변수명을 가지도록 수정하겠습니다
감사합니다!

seokhongkim commented on README.md:
````diff
@@ -132,4 +132,155 @@ npm run dev
+  ```
+
+- crontab 설정
+    - */60 * * * * /home/ubuntu/nextstep/script/deploy.sh step3 >> /home/ubuntu/nextstep/log/deploy.log 2>&1
````
crontab 설정 잘 해주셨습니다. 👍 
주기적으로 배포 시도 및 동일한 revision 일 때 배포하지 않는 것 확인 완료했습니다.

seokhongkim reviewed (changes requested):
보경님, 3단계 미션 잘 진행해주셨습니다! 👍 

질문 주신 부분에 대해 코멘트로 답변 남겼습니다.
추가로 배포 스크립트에 코멘트를 조금 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

ilmare-cbk commented:
> 보경님 안녕하세요! 질문 남겨주신 부분에 대한 답변 먼저 드립니다.
> 
> 1. [stackoverflow](https://stackoverflow.com/questions/30863012/hibernate-validation-on-server-startup-does-not-check-column-length)에 올라온 내용을 참고해보시면 도움이 될 것 같습니다. 실제로 org.hibernate.mapping.Table#validateColumns 의 구현을 보시면 type 과 컬럼이 없는 경우에 대한 체크만 있는 것을 보실 수 있습니다. :)
> 2. [askubuntu](https://askubuntu.com/questions/98782/how-to-run-an-alias-in-a-shell-script)에 올라온 내용을 참고하시면 shell script 내에서 alias 를 사용하실 수 있습니다.
>    alias 로 설정하는 코드를 `.bashrc` 등에 추가하신다면, alias 대신 환경변수로 등록하는 방법을 고민해보셔도 좋겠습니다. [spring boot external config](https://docs.spring.io/spring-boot/docs/1.5.6.RELEASE/reference/html/boot-features-external-config.html) 문서를 참고해보시면 10번에 OS environment variables 도 설정값으로 사용할 수 있는 것을 확인하실 수 있습니다. :)

오 감사합니다!!
공유해주신 링크 보면서 학습해보도록 하겠습니다 👍

ilmare-cbk committed:
docs(README.md): step3 요구사항 수정

- find_jar 함수를 통해 JAR_NAME 조회하도록 수정
- find_pid 함수를 통해 PID 조회하도록 수정
- master 변수명을 local로 변경

seokhongkim reviewed (approved):
보경님, 3단계 미션 코멘트도 잘 반영해주셨습니다! 👍 
3단계 PR도 머지하도록 하겠습니다.
코멘트 드린 내용 보시고, 관련하여 궁금하신 점이 있다면 슬랙 DM으로 편하게 말씀해주세요.

남은 교육 기간동안에도 지금처럼 열심히 학습하셔서 목표하신 성장과 성취 모두 이루시길 응원하겠습니다. 💪 
인프라 미션 수고 많으셨습니다!
